### PR TITLE
Convert scopes on upload

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Please see the commit messages for details.

### CODE REVIEW

File: model/src/main/kotlin/Project.kt [hunk 1]
```diff
@@ -111,6 +111,13 @@ data class Project(
         )
     }
 
+    init {
+        require(scopeDependencies == null || dependencyGraph == null) {
+            "Not both 'scopeDependencies' and 'dependencyGraph' may be set, as otherwise it is ambiguous which one " +
```
Comment (fviernau) [inline model/src/main/kotlin/Project.kt:116]: Maybe add a comma after both?

**Review by fviernau — DISMISSED** [review]

File: model/src/main/kotlin/Project.kt (hunk 1 above)
Comment (sschuberth) [reply model/src/main/kotlin/Project.kt:116]: Then you'd also need a comma after `'dependencyGraph'`. I think it's fine as-is.

File: cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt [hunk 2]
```diff
@@ -98,7 +100,7 @@ class UploadResultToPostgresCommand : CliktCommand(
         createConnection(postgresConfig).use { connection ->
             val query = "INSERT INTO ${postgresConfig.schema}.$tableName ($columnName) VALUES (to_json(?::json)::jsonb)"
 
-            val json = jsonMapper.writeValueAsString(ortResult).escapeNull()
+            val json = jsonMapper.writeValueAsString(ortResult.normalizeDependencies()).escapeNull()
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt:103]: Hmm, I'm not sure about the `normalizeDependencies()` name. Maybe better something like `convertGraphToScopes()`?

File: cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt [hunk 3]
```diff
@@ -147,5 +149,18 @@ class UploadResultToPostgresCommand : CliktCommand(
...
+    private fun AnalyzerRun.normalizeDependencies() =
+        copy(result = result.copy(projects = result.projects.mapTo(sortedSetOf()) { it.normalizeDependencies() }))
+
+    private fun Project.normalizeDependencies() =
+        dependencyGraph?.let {
+            copy(
+                scopeDependencies = it.createScopes(),
+                dependencyGraph = null
+            )
+        } ?: this
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt:163]: FYI, @oheger-bosch used to have this (by then unused) function in his PR:

https://github.com/oss-review-toolkit/ort/compare/5c5aa4555e6f80b325a5f58fdf444de5005d0664..a96a410847700e18423369864635a538c2623a1b#diff-6bcb52b873b6269f0794bf61bbb9f4d8fa1770d78f1650a613004d1ebd60c3ceL123-L136

Should we restore it for use from here?

File: cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt (hunk 2 above)
Comment (oheger-bosch) [reply cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt:103]: Agree with this comment. Which format is considered as normalized may change in future.

File: cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt (hunk 3 above)
Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt:163]: Sounds good to me, @oheger-bosch do you want to add this function? I would then update this PR afterwards.

Comment (oheger-bosch) [reply cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt:163]: @mnonnenmacher: Done in #3618.

Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/UploadResultToPostgresCommand.kt]: Thanks, I have updated this PR.

File: model/src/main/kotlin/Project.kt (hunk 1 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/Project.kt:116]: I have kept this as is to keep aligned with a similar message in `Provenance.kt` which I copied here.

File: model/src/main/kotlin/OrtResult.kt [hunk 4]
```diff
@@ -436,4 +436,18 @@ data class OrtResult(
      * Return true if and only if the given [id] denotes a [Project] contained in this [OrtResult].
      */
     fun isProject(id: Identifier): Boolean = getProject(id) != null
+
+    /**
+     * Resolves the scopes of all [Project]s in this [OrtResult] with [Project.withResolvedScopes].
+     */
+    fun withResolvedScopes(): OrtResult =
+        copy(
+            analyzer = analyzer?.let { analyzerRun ->
```
Comment (sschuberth) [inline model/src/main/kotlin/OrtResult.kt:445]: Could drop the `.let` like:

        copy(
            analyzer = analyzer?.copy(
                result = analyzer.result.copy(
                    projects = analyzer.result.projects.mapTo(sortedSetOf()) { it.withResolvedScopes() }
                )
            )
        )

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by fviernau — APPROVED** [review]

File: model/src/main/kotlin/OrtResult.kt (hunk 4 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/OrtResult.kt]: Done.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**